src/err/err_state.rs

```diff
@@ -69,25 +79,43 @@ impl PyErrState {
         // This process is safe because:
         // - Access is guaranteed not to be concurrent thanks to `Python` GIL token
         // - Write happens only once, and then never will change again.
-        // - State is set to None during the normalization process, so that a second
-        //   concurrent normalization attempt will panic before changing anything.
 
-        // FIXME: this needs to be rewritten to deal with free-threaded Python
-        // see https://github.com/PyO3/pyo3/issues/4584
+        // Guard against re-entrant normalization, because `Once` does not provide
+        // re-entrancy guarantees.
+        if let Some(thread) = self.normalizing_thread.lock().unwrap().as_ref() {
+            assert!(
+                !(*thread == std::thread::current().id()),
+                "Re-entrant normalization of PyErrState detected"
+            );
+        }
 
-        let state = unsafe {
-            (*self.inner.get())
-                .take()
-                .expect("Cannot normalize a PyErr while already normalizing it.")
-        };
+        self.normalized.call_once(|| {
+            self.normalizing_thread
+                .lock()
+                .unwrap()
+                .replace(std::thread::current().id());
+
+            // Safety: no other thread can access the inner value while we are normalizing it.
+            let state = unsafe {
+                (*self.inner.get())
+                    .take()
+                    .expect("Cannot normalize a PyErr while already normalizing it.")
+            };
 
-        unsafe {
-            let self_state = &mut *self.inner.get();
-            *self_state = Some(PyErrStateInner::Normalized(state.normalize(py)));
-            match self_state {
-                Some(PyErrStateInner::Normalized(n)) => n,
-                _ => unreachable!(),
+            let normalized_state = PyErrStateInner::Normalized(state.normalize(py));
```

Comment:
I think the only spot where there might be a deadlock is here, if `normalize` somehow leads to arbitrary Python code execution.

Is that possible? If not I think it deserves a comment explaining why.

Reply:
Great observation; I've added a wrapping call to `py.allow_threads` before potentially blocking on the `Once`, which I think avoids the deadlock (I pushed a test which did deadlock before that change).